### ISSUE TITLE
Cannot read binary file, and channel got closed

### ISSUE THREAD

**Issue by agusmakmun** [ronf/asyncssh#538]
![image](https://user-images.githubusercontent.com/7134451/214751169-44b69d20-9e5e-4a97-a5aa-9609f1265efb.png)

### Steps to reproduce the issue
1. Download this zip file [nginx-sniff-capture.zip](https://github.com/ronf/asyncssh/files/10506088/nginx-sniff-capture.zip)
2. Extract that binary file, named as `nginx-sniff-capture.pcap`
3. Try to read that file by using this asyncssh

In my case;

```python
connection = await asyncssh.connect(
    host=host,
    port=port,
    username=username,
    known_hosts=None,
    client_keys=settings.AWS_EC2_SSH_KEY_FILE,
)
```

I'm using Django & `AsyncWebsocketConsumer` for this: https://channels.readthedocs.io/en/latest/topics/consumers.html#asyncwebsocketconsumer

**Comment by ronf** [comment]
You don't show the code you are using to open the channel or read the data, but looking at the errors this is expected behavior.

If you want to read binary data (having the read call return `bytes` instead of `str`), you need to set `encoding=None` as one of the arguments when you open the channel or when you open the connection (in which case that will become the default for all channels on that connection).

**Comment by agusmakmun** [comment]
I see, I saw in my code is using decode `utf-8`.

```python
async def send(self, cmd):
    self.process.stdin.write(cmd.decode("utf-8"))
```

is it possible for `stdin.write` to process bytes? can you give me an example how to do that?

**Comment by ronf** [comment]
Yes - you do this by setting the encoding argument when opening the channel or connection, as I mention above. So, you could add `encoding=None` as an argument in the connect() call you show above, or you could add it to the call that creates the process object you are referencing in your `send()` call.

Once you've done that, calls to self.process.stdin.write() will expect a `bytes` argument, and calls to self.process.stdout.read() will return `bytes`.

**Comment by agusmakmun** [comment]
I tried to set `encoding=None` inside of the `asyncssh.connect` argument, but somehow the terminal browser won't opening.

```python
connection = await asyncssh.connect(
    host=host,
    port=port,
    username=username,
    known_hosts=None,
    client_keys=settings.AWS_EC2_SSH_KEY_FILE,
    encoding=None,
)
...
self.process = await self.connection.create_process(
    term_type="vt100", term_size=(tty_width, tty_height)
)
```

do I missed something here?

my `websocket.py`

```python
async def receive(self, text_data=None):
    """function that handle recieved message from client

    there are couple thigns it does, from validating connection to resize machine tty
    see below.
    --> validate connection, if already closed on server side. then stop
    --> parsed the message
    --> if message is heartbeat
        --> call check_connection.
    --> if message is resize terminal
        --> resize terminal using provider resize
    --> if message is message
        --> forward message to machine

    Parameters
    ----------
    text_data : byte, encoded str
        unparsed data from client.
    """
    if self.closed:
        return

    hearthbeat_message = "\0"
    if text_data == hearthbeat_message:
        self.logger.log(logger.debug, msg="ping lab machine")
        await self.check_connection()
        return

    text_data = text_data.encode("utf-8")
    try:
        json_data = json.loads(text_data)
        # handle invalid parsing, e.g:
        # returned as list, but expected is dict.
        if not isinstance(json_data, dict):
            json_data = {}
    except json.decoder.JSONDecodeError:
        json_data = {}

    try:
        # Resolved issue #13
        # Resize the terminal if payloads is given.
        status = json_data.get("status")
        tty_width = json_data.get("cols")
        tty_height = json_data.get("rows")
        if (status == 1) and (tty_width and tty_height):
            self.logger.log(logger.debug, msg="resizing terminal")
            await self.lab.resize_tty(tty_width=tty_width, tty_height=tty_height)
        else:
            await self.lab.send(text_data)

        logger.debug("## Message Received: " + str(text_data))
    except Exception:
        self.logger.log(
            logger.info, msg="send message to container cancelled, websocket closed"
        )
        await self.check_connection()
```

and my `lab.py`

```python
async def stream(self, on_recieve_callback: Callable):
    # send newline for xtermjs welcome text
    await on_recieve_callback("\n")

    # stream ssh data
    while self.running:
        line = await self.process.stdout.read(1)
        message = line if line else None
        await on_recieve_callback(message)

    self.logger.log(logger.debug, msg="stream finished")

async def send(self, cmd):
    self.process.stdin.write(cmd.decode("utf-8"))
```

**Comment by ronf** [comment]
In your send function, you should remove the `.decode('utf-8')`, as the write() call would now be expecting a `bytes` value.

In the stream function, the "line" value would now be `bytes` as well. Will the on_receive_callback() be ok with that? I see you're also calling that with "\n", which is a string. If the receive callback takes either `bytes` or `str`, that might be ok, but otherwise you should be consistent about always using `bytes`.

If the receive callback can only accept a `str`, that's going to be a problem if you expect to receive binary data on stdout.

Also, I noticed you are reading 1 byte here, which would be somewhat inefficient. You might want to think about using a larger value there. It will still return immediately with a smaller amount of data if less than that number of bytes is available.
